### PR TITLE
Converted castle_db.txt to YAML

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: N/A

* **Server Mode**: Both

* **Description of Pull Request**: 

  * Converts the Guild Castles Database into YAML.
  * Includes CSV2YAML converter.

### CODE REVIEW

File: src/common/mmo.hpp [hunk 1]
```diff
@@ -721,9 +722,9 @@ struct guild {
 
 struct guild_castle {
 	int castle_id;
-	int mapindex;
-	char castle_name[NAME_LENGTH];
-	char castle_event[EVENT_NAME_LENGTH];
+	uint16 mapindex;
+	std::string castle_name,
+		castle_event;
```
Comment (Lemongrass3110) [inline src/common/mmo.hpp:727]: I am not sure if this works.
This struct is used in inter-server communication as well. I am not sure if std::string is getting serialized and deserialized properly.
Did you test this?

File: src/map/guild.cpp [hunk 2]
```diff
@@ -269,25 +270,72 @@ bool guild_check_skill_require( struct guild *g, uint16 id ){
...
+		gc->castle_id = castle_id;
 	}
+
+	if (this->nodeExists(node, "Map")) {
+		std::string map_name;
+
+		if (!this->asString(node, "Map", map_name))
+			return 0;
+
+		uint16 mapindex = mapindex_name2id(map_name.c_str());
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:300]: By using this function missing maps are already getting reported. For a silent call you need to call it with mapindex_name2idx(mapname, nullptr)

Comment (Lemongrass3110) [reply src/map/guild.cpp:300]: Also this function returns the mapid and not the mapindex. So variable naming is bad or maybe even wrong logic?

File: src/map/guild.cpp [hunk 3]
```diff
@@ -269,25 +270,72 @@ bool guild_check_skill_require( struct guild *g, uint16 id ){
...
+		gc->castle_name = castle_name;
+	}
+
+	if (this->nodeExists(node, "NPC")) {
+		std::string npc_name;
+
+		if (!this->asString(node, "NPC", npc_name))
+			return 0;
+
+		if (npc_name.size() > EVENT_NAME_LENGTH) {
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:326]: I think what you want to check is NPC_NAME_LENGTH?
Also the error message below should be changed.
Unless you already want to build the full event name here and you forgot to append `::OnGuildBreak` here.

File: src/map/guild.cpp [hunk 4]
```diff
@@ -2066,20 +2096,17 @@ int guild_break(struct map_session_data *sd,char *name) {
  * from char-server.
  */
 void guild_castle_map_init(void) {
-	int num = db_size(castle_db);
```
Comment (Lemongrass3110) [inline src/map/guild.cpp:2069]: Please change this whole dirty function...
C++ iterate over it and build a vector with the ids that need to be requested.
Thank you!

File: src/map/guild.hpp [hunk 5]
```diff
@@ -119,4 +116,19 @@ void guild_retrieveitembound(uint32 char_id,uint32 account_id,int guild_id);
...
+class CastleDatabase : public TypesafeYamlDatabase <int32, guild_castle> {
+public:
+	CastleDatabase() : TypesafeYamlDatabase("CASTLE_DB", 1) {
+
+	}
+
+	const std::string getDefaultLocation();
+	uint64 parseBodyNode(const YAML::Node &node);
+
+	std::shared_ptr<guild_castle> guild_mapname2gc(const char* mapname);
```
Comment (Lemongrass3110) [inline src/map/guild.hpp:128]:
```suggestion
	std::shared_ptr<guild_castle> mapname2gc(const char* mapname);
```

File: src/map/guild.hpp [hunk 6]
```diff
@@ -119,4 +116,19 @@ void guild_retrieveitembound(uint32 char_id,uint32 account_id,int guild_id);
...
+	std::shared_ptr<guild_castle> guild_mapindex2gc(int16 mapindex);
```
Comment (Lemongrass3110) [inline src/map/guild.hpp:129]:
```suggestion
	std::shared_ptr<guild_castle> mapindex2gc(int16 mapindex);
```

File: src/map/mob.cpp [hunk 7]
```diff
@@ -827,7 +827,7 @@ int mob_spawn_guardian(const char* mapname, int16 x, int16 y, const char* mobnam
 	struct mob_data *md=nullptr;
 	struct spawn_data data;
 	struct guild *g=nullptr;
-	struct guild_castle *gc;
+	std::shared_ptr<guild_castle> gc = nullptr;
```
Comment (Lemongrass3110) [inline src/map/mob.cpp:830]: Move declaration down instead?

**Comment by Lemongrass3110** [comment]
Btw I know it does not make sense for you now, but can you please put the castle db into pre-re and re (duplicate) when you are done?
I will introduce a V2 for renewal later on. ;-)

File: src/common/mmo.hpp (hunk 1 above)
Comment (Atemo) [reply src/common/mmo.hpp:727]: castle_name and castle_event are not in int_guild.cpp nor in the guild_castle table (handled by int_guild.cpp)
```
CREATE TABLE IF NOT EXISTS `guild_castle` (
  `castle_id` int(11) unsigned NOT NULL default '0',
  `guild_id` int(11) unsigned NOT NULL default '0',
  `economy` int(11) unsigned NOT NULL default '0',
  `defense` int(11) unsigned NOT NULL default '0',
  `triggerE` int(11) unsigned NOT NULL default '0',
  `triggerD` int(11) unsigned NOT NULL default '0',
  `nextTime` int(11) unsigned NOT NULL default '0',
  `payTime` int(11) unsigned NOT NULL default '0',
  `createTime` int(11) unsigned NOT NULL default '0',
  `visibleC` int(11) unsigned NOT NULL default '0',
  `visibleG0` int(11) unsigned NOT NULL default '0',
  `visibleG1` int(11) unsigned NOT NULL default '0',
  `visibleG2` int(11) unsigned NOT NULL default '0',
  `visibleG3` int(11) unsigned NOT NULL default '0',
  `visibleG4` int(11) unsigned NOT NULL default '0',
  `visibleG5` int(11) unsigned NOT NULL default '0',
  `visibleG6` int(11) unsigned NOT NULL default '0',
  `visibleG7` int(11) unsigned NOT NULL default '0',
  PRIMARY KEY  (`castle_id`),
  KEY `guild_id` (`guild_id`)
) ENGINE=MyISAM;
```
so there are only used for the mapserver.
I made some tests on local, I have no issue yet

File: src/map/guild.cpp (hunk 3 above)
Comment (Atemo) [reply src/map/guild.cpp:326]: Ha true! I forgot half of the string

File: src/common/mmo.hpp (hunk 1 above)
Comment (Lemongrass3110) [reply src/common/mmo.hpp:727]: Talking about this:
```
int intif_parse_GuildCastleDataLoad(int fd)
{
	return guild_castledataloadack(RFIFOW(fd,2), (struct guild_castle *)RFIFOP(fd,4));
}
```

Comment (Atemo) [reply src/common/mmo.hpp]: I just reverted the changes in mmo

File: src/map/guild.cpp (hunk 4 above)
Comment (Atemo) [reply src/map/guild.cpp:2069]: The function is using some packets sent to charserver and I have some issue with that. I can't fix the issues on my local. Can you commit the changes when you have time?

Comment (Lemongrass3110) [reply src/map/guild.cpp:2069]: Done! :)

**Review by Lemongrass3110 — APPROVED** [review]

**Pull request merged**